snapshot-labs/snapshot-sidekick
# feat: return cache generation progress

## 🧿 Current issues / What's wrong ?

For proposal with a huge votes count, cache generation may take a while, and may take hours to complete.

There is currently no way to track the progress of a generation taking hours, so we don't know if the job is stuck somewhere, or still working.

## 💊 Fixes / Solution

Show the job progress status in %, in the logs, as well as return it when fetching a cache file which is pending generation.

## 🚧 Changes

- [refactor] Refactor queue.ts, to use the `queue` interface to add jobs to it, instead of manipulating the Set object directly
- [new] VotesReport object now return the generation progress status, by calling `generationProgress` property
- [refactor] In the queues, save the votesReport in a map, so we can retrieve their progress on demand
- [change] When retrieving a cache file, it no longer return a JSON-RPC error message when the cache is pending generation. It will instead return a JSON-RPC success response, with the progress percentage in the `result` field. Status code is still the same `202`. 

## 🛠️ Tests

- Start `yarn dev`
- Call `curl -X POST localhost:3005/api/votes/0x5280241b4ccc9b7c5088e657a714d28fa89bd5305a1ff0abf0736438c446ae98 -H "Content-Type: application/json"`
- Look at the yarn dev console, it should show the job progress every 15s (excluding the first one)

You can decrease the queue polling time (e.g.3s) in lib/queue.ts for faster test.

- When sending the curl request again, the `result` field should display the progress percentage.
